server/src/watchdog.js

```diff
@@ -0,0 +1,141 @@
+const config = require("./config").getProperties();
+const mozlog = require("./logging").mozlog("server");
+const db = require("./db");
+const { Shot } = require("./servershot");
+const util = require("util");
+const { URL } = require("url");
+const fetch = require("node-fetch");
+const Hawk = require("hawk");
+const FormData = require("form-data");
+const genUuid = require("nodify-uuid");
+
+const isEnabled = config.watchdog.enable;
+const isProperlyConfigured = validateConfiguration();
+
+function validateConfiguration() {
+  if (!isEnabled) {
+    // Do no warn about config values if Watchdog is not enabled.
+    return false;
+  }
+
+  let isConfigValid = true;
+
+  if (!config.watchdog.id || !config.watchdog.key || !config.watchdog.algorithm) {
+    mozlog.warn("watchdog-invalid-config", {msg: "Watchdog credentials configuration values are missing."});
+    isConfigValid = false;
+  }
+
+  if (!config.watchdog.submissionUrl) {
+    mozlog.warn("watchdog-invalid-config", {msg: "Watchdog submission URL configuration value is missing."});
+    isConfigValid = false;
+  } else {
+    try {
+      new URL(config.watchdog.submissionUrl);
+    } catch (e) {
+      mozlog.warn("watchdog-invalid-config", {msg: "Watchdog submission URL is not a valid URL.", err: e});
+      isConfigValid = false;
+    }
+  }
+
+  return isConfigValid;
+}
+
+let positiveEmailList;
+if (config.watchdog.positiveEmail) {
+  positiveEmailList = config.watchdog.positiveEmail.split(";");
+}
+
+let denominator = config.watchdog.throttleDenominator;
+if (denominator <= 1) {
+  denominator = 1;
+}
+let odometer = 0;
+
+function shouldSubmitShot() {
+  if (!isEnabled || !isProperlyConfigured) {
+    return false;
+  }
+  if (denominator === 1) {
+    return true;
+  }
+
+  const shouldSubmit = odometer === 0;
+  odometer = (odometer + 1) % denominator;
+
+  return shouldSubmit;
+}
+
+function getSubmissionId() {
+  return db.select(
+    "SELECT nextval(pg_get_serial_sequence('watchdog_submissions', 'id')) AS new_id",
+  ).then(rows => {
+    if (rows.length) {
+      return rows[0].new_id;
+    }
+    throw new Error("Unable to retrieve a new Watchdog submission id.");
+  });
+}
+
+const credentials = {
+  id: config.watchdog.id,
+  key: config.watchdog.key,
+  algorithm: config.watchdog.algorithm
+};
+
+function submit(shot) {
+  if (!shouldSubmitShot()) {
+    return;
+  }
+
+  const imageUrl = shot.getClip(shot.clipNames()[0]).image.url;
+  const imageId = imageUrl.split("/").pop();
+  const form = new FormData();
+
+  Shot.getRawBytesForClip(
```

Comment:
This would all be a lot nicer with async/await... though I suppose there's coverage concerns. Seems a shame, though.

Reply:
Do people look at or find those coverage reports helpful?

Reply:
>  Do people look at or find those coverage reports helpful?

I don't, but we could make test coverage a code quality theme for Q3 or Q4, and then it'd be a useful first step.

Reply:
I don't look at them personally. We're kind of catching up with tests, but I don't think we were in a bad place before, which is to say we aren't catching up with *quality*. So... I don't know. But I'm also leaning away from Screenshots, so my opinion on the matter isn't as important as y'all's.